File: ovn0/uppg3.py
# Nisse Nilsson, grudat18 uppg 0.3

class Stack:
	"""This class implements a stack of objects."""

	def __init__(self):
		"""Construct an empty stack."""
		# _data[] holds the stack items.
		self._data = []

	def push(self, item):
		"""Add x to the top of the stack."""
		self._data += [item]

	def pop(self):
		"""Remove and return the top element of the stack.

		It is a run-time error to call pop on an empty stack.
		"""
		return self._data.pop()

	def __len__(self):
		"""Return the number of elements in the stack."""
		return len(self._data)

# Unit test
def main():
	s = Stack() # stack of strings
	s.push("world!")
	s.push("Hello, ")
	assert len(s) == 2

	res = ""
	exp = "Hello, world!"
	while len(s) > 0:
		res += s.pop()
	assert res == exp

if __name__ == '__main__': main()

File: ovn1/uppg3.go
// Nisse Nilsson, grudat18 uppg 1.3

// This package implements a stack of strings.
package main

import "fmt"

// The zero value for Stack is an empty stack ready to use.
type Stack struct {
	data []string
}

// Push adds x to the top of the stack.
func (s *Stack) Push(x string) {
	s.data = append(s.data, x)
}

// Pop removes and returns the top element of the stack.
// It’s a run-time error to call Pop on an empty stack.
func (s *Stack) Pop() string {
	n := len(s.data) - 1
	res := s.data[n]
	s.data[n] = "" // to avoid memory leak
	s.data = s.data[:n]
	return res
}

// Size returns the number of elements in the stack.
func (s *Stack) Size() int {
	return len(s.data)
}

// Unit test
func main() {
	var s Stack
	s.Push("world!")
	s.Push("Hello, ")
	if size := s.Size(); size != 2 {
		fmt.Println("got", size, "expected", 2)
	}

	res := ""
	exp := "Hello, world!"
	for s.Size() > 0 {
		res += s.Pop()
	}
	if res != exp {
		fmt.Println("got", res, "expected", exp)
	}
}

